### ISSUE TITLE
Hash table index

### ISSUE THREAD

**Issue by wojons** [rethinkdb/rethinkdb#2600]
I know this is not on the road map at all but want to put it in for now so people think about it. I would like hash table based indexing. there are some cases where the table or the data is really only looked up by some id and B-tree is fast but hash table is just so much faster and will also remove reason to use a lot of other databases when sharding uses hashed aslo.

**Comment by xkxx** [comment]
Any updates on this? Would using a hash index drastically improve performance (from an order of log n to constant time)?

**Comment by danielmewes** [comment]
@xkxx Not yet, sorry.
To give a bit of background: I think in practice the performance difference will be less than one might expect. A lot of processing time is actually needed for things like query parsing, converting results into JSON responses etc.  While all of that is O(1) for a constant-size point get, the O(log n) btree lookup  as of now is usually not the primary bottleneck. We could maybe get 10-20% more performance out of a hash index, but our main focus for now is on the other 80%.

**Comment by wojons** [comment]
@danielmewes 

I can see in a normal workload what your saying makes a lot of sense but what about when someone is doing a join btween two tables and the table getting joined is with hash index the join should take a lot less time to happen since the query its self was already parsed from the orginal

**Comment by danielmewes** [comment]
@wojons You are right, that makes sense. We first need to optimize `eq_join` though (at the moment it more or less does a `getAll` for each key).

**Comment by wojons** [comment]
@danielmewes I will leave it to you grand master
